CONTRIBUTING.md

```diff
@@ -0,0 +1,232 @@
+
+  The only exception to this rule is for [unhandled exceptions in background threads](#bgexceptions)
+
+- *Do not use exceptions for normal control flow*
+
+  Prefer to write code that avoids exceptions being thrown (e.g. the `TryXXX` pattern).  There are performance costs to using exceptions for control flow, and VFS4G code typically needs to handle errors where (and as soon as) they occur.
+   
+- *Provide the user with user-actionable messages whenever possible*
+
+  Don't tell a user what went wrong.  Help the user fix the problem.
+
+  Example:
+  > "You can only specify --hydrate if the repository is mounted. Run 'gvfs mount' and try again."
+
+## Background Threads
+
+- *Avoid using the thread pool (and avoid using async)*
+
+  `HttpRequestor.SendRequest` makes a [blocking call](https://github.com/Microsoft/VFSForGit/blob/4baa37df6bde2c9a9e1917fc7ce5debd653777c0/GVFS/GVFS.Common/Http/HttpRequestor.cs#L135) to `HttpClient.SendAsync`.  That blocking call consumes a thread from the managed thread pool.  Until that design changes, the rest of VFS4G must avoid using the thread pool unless absolutely necessary.  If the thread pool is required, any long running tasks should be moved to a separate thread managed by VFS4G itself (see [GitMaintenanceQueue](https://github.com/Microsoft/VFSForGit/blob/4baa37df6bde2c9a9e1917fc7ce5debd653777c0/GVFS/GVFS.Common/Maintenance/GitMaintenanceQueue.cs#L19) for an example).
+
+  Long-running or blocking work scheduled on the managed thread pool can prevent the normal operation of VFS4G.  For example, it could prevent downloading file sizes, loose objects, or file contents in a timely fashion.
+
+- <a id="bgexceptions"></a>*Catch all exceptions on long-running tasks and background threads*
+
+  Wrap all code that runs in the background thread in a top-level `try/catch(Exception)`.  Any exceptions caught by this handler should be logged, and then force VFS4G to terminate with `Environment.Exit`.  It's not saft to allow VFS4G to continue to run after an exception that would have been unhandled in a long-running task (created with `TaskCreationOption.LongRunning`) or in a background thread.
+
+  An example of this pattern can be seen [here](https://github.com/Microsoft/VFSForGit/blob/4baa37df6bde2c9a9e1917fc7ce5debd653777c0/GVFS/GVFS.Virtualization/Background/BackgroundFileSystemTaskRunner.cs#L233) in `BackgroundFileSystemTaskRunner.ProcessBackgroundTasks`.
+
+## Coding Conventions
+
+- *Most C# coding style guidelines are covered by StyleCop*
+
+  Fix any StyleCop issues reported in changed code. When adding new projects to VFS4G, be sure that StyleCop is analyzing them as part of the build.
+
+- *Prefer explicit types to interfaces and implicitly typed variables*
+
+  Avoid the use of `var` (C#) and `auto` (C++), and prefer concrete/explicit types to interface (e.g. prefer `List` to `IList`).
```

Comment:
Why are you discouraging the use of interfaces? Seems to go against a few of the SOLID principles.